### PR TITLE
 Graceful custom resource storage teardown

### LINKED ISSUE #68508: Should support graceful deletion for CR storage
/kind bug
/sig api-machinery

another bug uncovered by #68405. in the extension server, we are removing old CR's storage when some significant changes (spec, some status) land in the corresponding CRD with the help of CRD informer. but when we're destroying the old CRD's storage, there are might be some remaining requests invoking etcd client from the old storage. so a racing happens when the CRD's informer trying to delete/close the old storage together with the old etcd client and the running handler trying to call REST interface via that client. as a result, the handler will fail with a `context canceled` or `grpc: the client connection is closing` which is the reason #68405 is flaky. 

in a word, __updating CRD will fail all processing CR requests__. 

@sttts proposes to make use of the [apimachinery WaitGroup](https://github.com/kubernetes/kubernetes/pull/54849) for the storage and we only delete a storage when the all the running handler's finished. but we might left some problem with that:

- do we count the `WaitGroup` only in the handler ignoring other etcd client references like compactor?  
- ~how will this impact server's performance?~
- does it have to be a `WaitGroup`?
- what about some long running handlers? `watch` has a timeout itself tho, what about infinite handlers like `proxyport`?

otoh, we can add a cooldown period (minute-ish?) for older cr storage to postpone the actual deletion for that storage. surely, it's not stable but a cooldown will at least keep most of short-running handler from failing.

@deads2k @stttts

### PR DESCRIPTION
**Pull request by liggitt**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
Tracks in-flight requests for short-lived custom resource requests and waits to destroy backing storage until those in-flight requests are finished. Matches behavior for in-tree resources.

**Which issue(s) this PR fixes**:
Fixes #68508

**Does this PR introduce a user-facing change?**:
```release-note
Fixed a transient error API requests for custom resources could encounter while changes to the CustomResourceDefinition were being applied.
```

/sig api-machinery
/cc @jpbetz @sttts 
/priority important-soon

### CODE REVIEW

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go [hunk 1]
```diff
@@ -404,14 +413,38 @@ func (r *crdHandler) removeDeadStorage() {
...
+
+// Wait up to a minute for requests to drain, then tear down storage
+func tearDown(oldInfo *crdInfo) {
+	requestsDrained := make(chan struct{})
+	go func() {
+		defer close(requestsDrained)
+		// Allow time for in-flight requests with a handle to the old info to register themselves
+		time.Sleep(time.Second)
+		// Wait for in-flight requests to drain
+		oldInfo.waitGroup.Wait()
```
Comment (jpbetz) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go]: Too bad we can't context cancel the in-flight requests.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go [hunk 2]
```diff
@@ -247,18 +254,19 @@ func (r *crdHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
...
 	default:
 		http.Error(w, "the server could not find the requested resource", http.StatusNotFound)
 	}
 
-	if handler != nil {
-		handler = metrics.InstrumentHandlerFunc(verb, requestInfo.APIGroup, requestInfo.APIVersion, resource, subresource, scope, metrics.APIServerComponent, handler)
-		handler(w, req)
+	if handlerFunc != nil {
+		handlerFunc = metrics.InstrumentHandlerFunc(verb, requestInfo.APIGroup, requestInfo.APIVersion, resource, subresource, scope, metrics.APIServerComponent, handlerFunc)
+		handler := genericfilters.WithWaitGroup(handlerFunc, longRunningFilter, crdInfo.waitGroup)
```
Comment (jpbetz) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go]: Great! This is much cleaner than the decorator.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go [hunk 3]
```diff
@@ -169,6 +174,8 @@ func NewCustomResourceDefinitionHandler(
 	return ret, nil
 }
 
+var longRunningFilter = genericfilters.BasicLongRunningRequestCheck(sets.NewString("watch"), sets.NewString())
```
Comment (jpbetz) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go]: Comment here on why tearing down rest storage in the presence of watches is safe?

**Review by jpbetz — COMMENTED** [review]
Looks good. Minor comments only.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go (hunk 1 above)
Comment (liggitt) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go]: they'll get timed out at the configured request timeout by https://github.com/kubernetes/kubernetes/blob/0252a323423080a3a230ab3053fdaaca8a972eb1/staging/src/k8s.io/apiserver/pkg/server/config.go#L552

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go (hunk 3 above)
Comment (liggitt) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go]: added a comment. for what it's worth, I have an expanded version of this test locally that also starts watches to make sure they get terminated cleanly. that recreates the watch hang bug. once this merges, I'll get that in a PR for use in verifying the watch hang when we fix that

**Comment by liggitt** [comment]
comments addressed

**Review by yue9944882 — COMMENTED** [review]
[https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiextensions-apiserver/test/integration/versioning_test.go#L80](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiextensions-apiserver/test/integration/versioning_test.go#L80)

let's remove this line?

**Comment by liggitt** [comment]
> https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiextensions-apiserver/test/integration/versioning_test.go#L80
> 
> let's remove this line?

no, we still need to give time for the CRD update to trigger replacing the handler. This just fixes spurious errors, it doesn't make the handler swap instantaneous.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go (hunk 1 above)
Comment (jpbetz) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go]: ah, nice

**Pull request merged**